design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,194 @@
+
+## Getting the information stored in the error info
+
+As mentioned earlier, projections should expose the details stored in the xlang error info as part of the function used in the language they project to log errors / exceptions (i.e. ToString on an exception).  This will allow consumers to get a full picture of the failure when they log an error using the natural and familiar code they would typically use.  To achieve this, xlang provides the IXlangErrorInfo interface.
+
+```
+interface IXLangErrorInfo
+{
+	void GetError(xlang_result* error)
+	void GetMessage(xlang_string* message)
+	void GetLanguageError(xlang_string* error)
+	void GetStack(xlang_string* stack)
+	void GetLanguageIdentifier(xlang_string* identifier)
+	void GetLanguageInformation(IUnknown* languageInformation)
+	void GetLanguageInformationAsString(xlang_string* languageInformation)
+	void GetPropagatedError(IXlangErrorInfo** propagatedError)
+}
+```
+
+Most of the functions provided in this interface are self-explanatory.  But there are a few which should be called out.  There are 2 variants of the function used to get the additional language specific information (GetLanguageInformation and GetLanguageInformationAsString).  The first function is expected to be used in scenarios where the projection itself wants to retrieve an object they stored.  The second function is expected to be used in scenarios where a projection is exposing the details in the error info object as part of the language function used to log errors.  This function will provide the information as a string if the COM object stored implements the IXlangErrorValue interface.  If it doesn't or no COM object was stored, it would just provide an empty string.  Another function to call out is the GetPropagatedError function.  This function will allow to traverse the diagnostic information from each language which the error propagated through.  For all the propagations, the error and message functions will provide the same values.  The remaining functions will provide the respective information the language captured.
+
+## Error free functions
+
+Some languages have a concept of error free functions (i.e. noexcept in C++).  These functions will not be projected as error free because errors can occur while translating the call from the other language and while translating the result back (i.e. during activation of the class for the function).  But if there was any error during the execution of the function itself, the language itself may take actions based on what it does for errors in error free functions (i.e. fail fast for noexcept).
```

Comment:
Feels like this is in the "guidance to projection authors" section.  Activation can fail, but are there other 'this function really really can't fail without it being end-of-days' markers? Is there a MIDL3 marker for 'this method cannot fail' that could be projected into noexcept or per-language equivalents?

Reply:
We do have a MIDL3 marker for this, but apparently not all projections recognize it.  I updated this section such that projections that do recognize it can do so and gain potential optimizations.